File: leetcode-cn/5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#


# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expand(s, left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1

            return left + 1, right - 1

        n = len(s)
        left, right = 0, 0
        for i in range(n):
            left1, right1 = expand(s, i, i)
            left2, right2 = expand(s, i, i + 1)

            if right1 - left1 > right - left:
                right, left = right1, left1

            if right2 - left2 > right - left:
                right, left = right2, left2

        return s[left:right + 1]


# @lc code=end

File: leetcode-cn/C++/5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 */

// @lc code=start
#include <string>
#include <vector>
using namespace std;

class Solution
{
public:
    string longestPalindrome(string s)
    {
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n, false));

        for (int i = 0; i < n; i++)
        {
            dp[i][i] = true;
        }

        int start = 0;
        int end = 0;

        for (int len = 2; len <= n; len++)
        {
            for (int i = 0; i < n; i++)
            {
                int j = i + len - 1;
                if (j >= n)
                {
                    break;
                }

                if (s[i] == s[j])
                {
                    if (len == 2 || (i + 1 <= j - 1 && dp[i + 1][j - 1]))
                    {
                        dp[i][j] = true;
                    }
                }

                if (dp[i][j] && j - i > end - start)
                {
                    start = i;
                    end = j;
                }
            }
        }
        return s.substr(start, end - start + 1);
    }
};
// @lc code=end
